pb/compressed vector serialization

peterboncz opened the pull request:
Vector::Serialize() and Vector::Deserialize() now understand CONSTANT, SEQUENCE and DICTIONARY vectors, and do not flatten these; rather keep them compressed (writing an optional field vector_type that by absence is FLAT_VECTOR).

In the case of DICTIONARY vectors,  only the *used* part of a dictionary is serialized. Note that in storage, dictionaries are created on the RowGroup level and hence a dictionary is often larger than a vector, so serializing the DICTIONARY vector as-is would be larger than the flattened version. Also, RLE storage leads to dictionary vectors that only use a small contiguous subset of the dictionary entries, which generates this problem. By creating a subset of the dictionary, this is prevented. This PR aims to always do better in terms of byte-size than the previous approach.

Now the Serialize/Deserialize methods always attempt compression, at least on the latest serialization version, by using ShouldSerialize(5) aka v1.3.0 (or v1.2.1). But these new versions would have to be added still. **For the moment I have put in ShouldSerialize(1)** to trigger the code in CI - especially the DEBUG run does serialize/deserialize for all query results in the Verify() code. 

Both methods defer to the old code in new private methods resp. SerializeFlat/DeserializeFlat if there is no vector compression.

An alternative would be to name the methods SerializeCompressed() and DeserializeCompressed() and not rename the previous methods in SerializeFlat/DeserializeFlat. In that case, MotherDuck would be in charge  of deciding when to call SerializeCompressed() or not. That would also potentially allow to enrich the method with more parameters. I was specifically thinking about a Map of dictionary_ids/value-bitmaps. In a stateful connection, this would allow to incrementally send over dictionaries. Not sure whether that complexity would be worth it, though.

Commit:
make serialization respect compressed vectors

conserve constant, sequence and dictionary-vectors while serializing
(to save serialization space).

dictionary-vectors only serialize the used part of a dictionary.

Commit:
some debugging

Commit:
Merge branch 'pb/compressed-vector-serialization' of github.com:peterboncz/duckdb into pb/compressed-vector-serialization

Commit:
fix: dict vs &dict

Commit:
bring to the right serialization version

Commit:
back into voodoo-land with the irreproducible DDB make format

Commit:
DDB format-fix voodoo #2

Commit:
voodoo #3

Commit:
set ShouldSerialize(1) aka v.0.1.2 -- should become ShouldSerialize(5) before merge

but this way the PR gets a CI run that triggers the new code

peterboncz commented:
OK, this PR is passing all tests, including the DEBUG run. I had confirmed on my laptop with stderr printing that such a 'make unittest' run does produce a lot of CONSTANT, SEQUENCE and DICTIONARY (de)serialization.

Apart from any PR review comments, there is the question how to move forward with serialization. My personal interests would be to introduce it in v1.2.1, but if there are rules or reasons why it should be v1.3.0 that would certainly also work.

Mytherin commented on src/common/types/vector.cpp:
```diff
@@ -1463,14 +1539,15 @@ void Vector::Verify(Vector &vector_p, const SelectionVector &sel_p, idx_t count)
-		sel = &owned_sel;
-		vector = &child;
-		vtype = vector->GetVectorType();
+		if (child.GetVectorType() == VectorType::DICTIONARY_VECTOR) {
```
This has not been supported in the past - why should it be supported now?

peterboncz replied:
good catch, this was just plain wrong

Mytherin commented on src/common/types/vector.cpp:
```diff
@@ -1373,6 +1373,82 @@ void Vector::Deserialize(Deserializer &deserializer, idx_t count) {
+
+void Vector::Deserialize(Deserializer &deserializer, idx_t count) {
+	const auto vtype = // older versions that only supported flat vectors did not serialize vector_type,
+	    deserializer.ReadPropertyWithExplicitDefault<VectorType>(200, "vector_type", VectorType::FLAT_VECTOR);
```
Serialization identifiers should be incremental, we are now reading 200 first, and then potentially reading 100 after if the vector is a flat vector. 

Perhaps we can serialize the vector type at `99` instead?

peterboncz replied:
done

peterboncz replied:
I changed the method structure because in trying to develop testing for this at MotherDuck, in its hybrid_core CI, it we have the same DDB client and server version, so to trigger (no) compression an explicit parameter is useful. This also led me to inline the code inside the Serialzie method, but it could of course be split out. 

Brutale vraag: if I would put the default of the `compressed_serialization = false` and then lower to `ShouldSerialize(4)`, might this then potentially enter 1.2.1? In main/1.3 it would remain `= true` and `(5)`.

Mytherin replied:
I think this is better suited for 1.3

Mytherin reviewed (commented):
Thanks for the PR! Looks good - some comments:

Commit:
Merge remote-tracking branch 'origin/master' into pb/compressed-vector-serialization

Commit:
- work on review feedback
- add boolean parameter to Vector::Serialize and DataChunk::Serialize
  (this allows MotherDuck to test the effects of compressed_serialization)

peterboncz commented on src/common/types/vector.cpp:
```diff
@@ -1373,6 +1373,82 @@ void Vector::Deserialize(Deserializer &deserializer, idx_t count) {
 
+void Vector::Serialize(Serializer &serializer, idx_t count) {
+	// serialize compressed vectors to save space, but skip this if serializing into older versions
+	if (serializer.ShouldSerialize(1)) {
```
done

I did change the PR in that I added an optional boolean compress_vectors = true parameter, because for testing purposes (in MD) this allows to switch off vector compression (via DataChunk::Serialize that also gets the optional parameter)

Commit:
DDB format fix

Mytherin commented on src/include/duckdb/common/types/data_chunk.hpp:
```diff
@@ -143,7 +143,7 @@ class DataChunk {
 	DUCKDB_API void Reset();
 
-	DUCKDB_API void Serialize(Serializer &serializer) const;
+	DUCKDB_API void Serialize(Serializer &serializer, bool compressed_serialization = true) const;
```
Can we remove these parameters, or do they still need to be here?

Mytherin commented:
Thanks!
